File: SHARED/taskUtil.ts
const AIDEVS_BASE_URL = "https://tasks.aidevs.pl/"

export const getTask = async (taskName: string) => {
  const apiKey = process.env.AIDEVS_API_KEY;
  const headers = {
    'Content-Type': 'application/json',
    // add other headers if needed
  };

  let url = `${AIDEVS_BASE_URL}token/${taskName}`;
  let data = { apikey: apiKey };

  let response = await fetch(url, {
    method: 'POST',
    headers: headers,
    body: JSON.stringify(data)
  });

  let parsedResponse: any = await response.json();
  const token = parsedResponse.token;
  console.log('Token:', token)

  url = `${AIDEVS_BASE_URL}task/${token}`;
  response = await fetch(url, { method: 'POST' });
  parsedResponse = await response.json();

  return { theTask: parsedResponse.question, taskToken: token };
};

export const submitTask = async (answer: string, token: string) => {
  const url = `${AIDEVS_BASE_URL}answer/${token}`;
  const data = { answer: answer };
  const headers = {
    'Content-Type': 'application/json',
    // add other headers if needed
  };

  const response = await fetch(url, {
    method: 'POST',
    headers: headers,
    body: JSON.stringify(data)
  });

  console.log("submit result:", response);
  const parsedResponse = await response.json();
  console.log('submit response:', parsedResponse);

  return parsedResponse;
};

export async function getOaiAnswer(userPrompt: string, systemPrompt?: string, functions?: any) {
  const BASE_URL = "https://api.openai.com/v1/chat/completions";
  const OPENAI_API_KEY = process.env.OPENAI_API_KEY;
  const url = BASE_URL;
  const headers = { 'Authorization': `Bearer ${OPENAI_API_KEY}`, 'Content-Type': 'application/json' };

  let messages: any[] = [];
  if (systemPrompt !== null) {
    messages = [{ role: "system", content: systemPrompt }];
  }
  messages.push({ role: "user", content: userPrompt });

  const data = { model: "gpt-4", messages: messages };
  if (functions) {
    //@ts-ignore
    data['functions'] = functions;
    //@ts-ignore
    data['function_call'] = "auto";
  }

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: headers,
      body: JSON.stringify(data)
    });
    const responseData: any = await response.json();
    const responseMessage = responseData.choices[0].message;
    if (responseMessage['function_call']) {
      const fc = responseMessage['function_call']
      const functionName = fc['name']
      const functionArgs = JSON.parse(fc['arguments'])
      return {
        tool: functionName,
        arguments: functionArgs
      }
    } else {
      const resp = responseMessage.content;
      return resp;
    }
  } catch (error) {
    console.error(error);
  }
}

File: PYTHON_2/SHARED/task_util.py
import requests
import json

import os
from dotenv import load_dotenv
load_dotenv()

import sys

BASE_URL = "https://tasks.aidevs.pl/"
HEADERS = {'Content-Type': 'application/json'}
TASK_MODERATION = "moderation"
TASK_HELLO = "helloapi"

def submit_task(answer, token):
  url = BASE_URL + "/answer/" + token
  data = {"answer": answer }
  res = requests.post(url, data = json.dumps(data), headers = HEADERS )
  print("submit result:", res)
  response = json.loads(res.text)
  print('submit response:', response)


def get_task(task_name):
  url = BASE_URL+ "token/" + task_name
  data = {"apikey": os.getenv("AIDEVS_API_KEY")}

  response = requests.post(url, data=json.dumps(data), headers=HEADERS)
  parsed_response = json.loads(response.text)
  
  token = parsed_response['token']

  url = BASE_URL + "/task/" + token
  response = requests.post(url)
  parsed_response = json.loads(response.text)
  return parsed_response, token

if __name__ == "__main__":
  load_dotenv()
  
  if len(sys.argv) > 1:
    task_name = sys.argv[1]
  else:
    task_name = TASK_HELLO
    
  get_task(task_name)


def get_oai_answer(user_prompt, system_prompt = None):
  
  BASE_URL = "https://api.openai.com/v1/chat/completions"
  OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
  url = BASE_URL 
  headers = {"Authorization": "Bearer " + OPENAI_API_KEY}
  
  messages = []
  if system_prompt != None:
    messages = [{"role": "system", "content": system_prompt}]
  messages.append({"role": "user", "content": user_prompt})
  
  data = { "model": "gpt-4", "messages": messages }
  
  #response = requests.post(url, data=json.dumps(data), headers=headers)
  response = requests.post(url, json=data, headers=headers)#,verify=false
  
  # Parse JSON response
  parsed_response = json.loads(response.text)
  resp = parsed_response['choices'][0]['message']['content']
  
  return resp

def get_oai_answer_from_image(user_prompt, image_url):
  
  OAI_URL = "https://api.openai.com/v1/chat/completions"
  OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
  headers = {"Authorization": "Bearer " + OPENAI_API_KEY}
  
  messages=[
    {
      "role": "user",
      "content": [
        {"type": "text", "text": user_prompt},
        {
          "type": "image_url",
          "image_url": {
            "url": image_url
          },
        },
      ],
    }
  ]
  
  data = { "model": "gpt-4o", "messages": messages }
  
  response = requests.post(OAI_URL, json=data, headers=headers)
  
  parsed_response = json.loads(response.text)
  #print(parsed_response)
  resp = parsed_response['choices'][0]['message']['content']
  print(resp)
  return resp
